### ISSUE TITLE
Support Raw Sockets

### ISSUE THREAD

**Issue by chrisdutz** [netty/netty#7552]
We are using Netty 4 quite intensely in the new Apache PLC4X (Incubating) project in order to implement industrial communication protocols. This has been working nicely for TCP based protocols, but now we would like to start implementing protocols based on IP, but not TCP or UDP. 

We think it should be possible to implement such a RawSocketChannel based on the RockSaw project (Apache 2.0 Licensed) (https://www.savarese.com/software/rocksaw/ with maven usable artifacts from here https://github.com/mlaccetti/rocksaw)  and have started implementing such a thing inside the Apache PLC4X (incubating) project, but I think it would be a much better fit to have such code here.

As we are not that experienced with the inner workings of Netty 4, I guess such a joint effort could be beneficial for both projects.

**Comment by Scottmitch** [comment]
This sounds interesting. I'm not sure if any existing core developers will have cycles to take this on near term, but if you are willing to maintain the contribution PRs are always welcome. Note we have a few JNI based transports already in Netty (e.g. [Epoll](https://github.com/netty/netty/tree/4.1/transport-native-epoll) and [KQueue](https://github.com/netty/netty/tree/4.1/transport-native-kqueue)) and these both share some common [unix JNI utilities](https://github.com/netty/netty/tree/4.1/transport-native-unix-common). For the near term I would recommend using these locations as reference points to get started.

**Comment by chrisdutz** [comment]
Ok. This was what I found out too. Also that kqueue seems to be a copy of epoll (out the other way around). Right now I'm working myself through the details, trying to create a Netty transport in the Apache PLC4X (incubating) project. As soon as that's done I think I'll move that to a pull request for Netty and switch to the Netty version as soon as that's become a part of Netty.

**Comment by Scottmitch** [comment]
> Also that kqueue seems to be a copy of epoll (out the other way around).

Yes there is a large amount of overlap. We have not yet attempted to consolidate more code between these two transports because there were "enough" actual differences and unfortunately the class hierarchy is meaningful in both cases.

For raw sockets would you need to implement an entirely new selector mechanisms (and associated transport/event loop), or would this just be another channel type in the kqueue/epoll transports?

**Comment by chrisdutz** [comment]
I have no idea on what I would need. Right now I'm working on this with very little as I'm still on holidays, but when I'm back at work next week, I'll greatly increase the efforts.

Right now I'm trying to find out what the different parts are, what they do, why they are needed and what I would need. Unfortunately there is very little documentation so it's a little tricky.

Right now I think all I would need, is to replace the actual connecting, disconnecting, writing and reading of epoll or kqueue with that of rocksaw.

**Comment by Scottmitch** [comment]
> Unfortunately there is very little documentation so it's a little tricky.

Yes this is not a very common area for folks to tread in. If you develop any useful documentation along the way, PRs are welcome.

> Right now I think all I would need, is to replace the actual connecting, disconnecting, writing and reading of epoll or kqueue with that of rocksaw.

I haven't investigated, but I'm hopeful raw socket will just be a new channel type in epoll/kqueue, and will not require its own full transport. I would expect some additions to `transport-native-unix-common` for the raw socket operations.

**Comment by normanmaurer** [comment]
@chrisdutz do you have any update or should we close this ?

**Comment by chrisdutz** [comment]
Still working on it. Has just recently become higher priority for us at the PLC4X project. However I am asking a little finding out what's essential and what's not

**Comment by F4k3rzZ** [comment]
@chrisdutz what's your status this? I had the same idea for a couple of days now. I am willing to help to develop this feature.

**Comment by JulianFeinauer** [comment]
Hi folks, I also jump in here. I'm also from the plc4x project and just hacked support for serial port communication. So I will try to move things forward here as well.
@F4k3rzZ I'll keep you updated with the progress so that you can join the work :) 

Julian

**Comment by JulianFeinauer** [comment]
Hi all, just a short update... We have a working implementation for raw sockets now in plc4x. 
Only drawback is that it is based on deprecated OioEventLoop but works flawlessly, yet.
